[euchangxian] iP

euchangxian opened the pull request:
#  🚀  Changoose  🚀 
> Best task manager ever - Mom
> Using Changoose changed my life for the better - Girlfriend ~~(real)~~

## 🎉 This is the latest release of Changoose v0.1! 🎉 
- Manage your **TODOS**!
- Manage your **DEADLINES**!
- Manage your **EVENTS**!

## 🤩 What's New? 🤩
1. **All-new** name! _No more_ old boring Duke! Say hi to ***Changoose***!
2. **Cutting-edge** ***Task Manager*** technology!
3. **State-of-the-art** intelligence! Changoose can understand ~~(almost)~~ anything you tell it!

## 🦾 Performance Enhancements  🦾
- [x] Upgraded Changoose's intelligence.
     - Now you can add tasks by typing `deadline feed goose \by today` and Changoose will understand you! 
- [x] Fed the 🐹hamsters🐹. Now Changoose is faster than ever!
- [ ] Completes your tasks for you (Coming soon)

## TODO
```Java
/**
* Completes the given task.
* 
* @param task The task to complete.
*/
public static void completeTask(Task task) {
  // TODO blank because I don't know what to put here.
  return;
}
```
### Here is the link to the code for Changoose!
[Changoose](https://github.com/euchangxian/ip)

![Maxwell Cat](https://media.tenor.com/9h6Db9JmjAwAAAAd/maxwell-cat.gif)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

euchangxian committed:
Rename chatbot to Changoose and add dividers

euchangxian committed:
Add Duke::startParse and Duke::echo

euchangxian committed:
Replace usage of System.out.println with echo()

This reduces repeated code.

euchangxian committed:
Add indentLeft() and indent echo's output

euchangxian committed:
Add botName and scanner as Duke instance variables

euchangxian committed:
Add getter for Duke botName

euchangxian committed:
Change startParse() and echo() to instance methods

euchangxian committed:
Add taskList field

euchangxian committed:
Update startParse() to add input into taskList

euchangxian committed:
Update echo() output text

euchangxian committed:
Update echo() to take in a prefix

This is done by adding an overloaded echo() method to maintain the
functionality of existing code.

euchangxian committed:
Add printTasks() method

euchangxian committed:
Add line breaks in between methods

For improved readablity

euchangxian committed:
Update echo() to handle a list of strings

euchangxian committed:
Fix printing of dividers between every task

euchangxian committed:
Add switch case to handle list command

euchangxian committed:
Add Task class

euchangxian committed:
Update fields and methods to use Task class

euchangxian committed:
Remove useless scanner field

euchangxian committed:
Add markTask() and unmarkTask()

euchangxian committed:
Add cases for mark and unmark

Update startParse() to first split by spaces to handle commands with
multiple words, like mark x or unmark x, where x is the index of the
command.

euchangxian committed:
Add getters for taskName and isDone in Task

euchangxian committed:
Extract print methods to OutputService

This is to prevent Duke class from getting too bloated.

euchangxian committed:
Remove Duke field in OutputService

euchangxian committed:
Change entry point of program to CliParserService

This is to prevent bloating of Duke class

euchangxian committed:
Add custom Exceptions

euchangxian committed:
Add TaskType enum

euchangxian committed:
Add variants of Tasks

euchangxian committed:
Add cases for variants of Task

euchangxian committed:
Move main method back to Duke

euchangxian committed:
Update text-ui-test test case

euchangxian committed:
Personalize error messages

euchangxian committed:
Add delete functionality

euchangxian committed:
Add getNumberOfTasks()

Also removed usage of `this` where relevant to maintain consistency.

euchangxian committed:
Change Task to abstract class

euchangxian committed:
Change certain Duke methods to return an Optional

euchangxian committed:
Add handleDelete()

This is to handle the printing of extra information for the delete
command.

euchangxian committed:
Refactor createTask()

Move printing of error message to CliParserService

euchangxian committed:
Fix dependency injection

euchangxian committed:
Implement Serializable interface for Task

This allows object persistence for Task by making it serializable.

euchangxian committed:
Add StorageService class

This class handles the read/write to files for object persistence,
specifically taskList.

euchangxian committed:
Change DukeException to a Checked Exception

Update method signature of relevant methods.

euchangxian committed:
Change DukeException to a Checked Exception

Update method signature of relevant methods.

euchangxian committed:
Update display message when there are no tasks left

euchangxian committed:
Add FileCorruptedException

euchangxian committed:
Update StorageService thrown exceptions

StorageService now throws DukeStorageException instead of DukeException
for more specificity and different erorr handling.

euchangxian committed:
Update gitignore to ignore storage file

euchangxian committed:
Remove caching of taskList

The added complexity is not worth it for the increase in performance,
because the assumption was that the taskList would contain at most 100
Tasks only.

euchangxian committed:
Fix taskList not loading from storage

euchangxian committed:
Merge branch 'branch-Level-7'
Add Tasks List persistence ability to Duke.

euchangxian committed:
Rework parseTaskCommand() and createTask()

deadline commands must have an end date as argument, prefixed with 'by',
while event commands must have a start date prefixed with 'from', end
date prefixed with 'to' as arguments.

euchangxian committed:
Add TimeUtilException

euchangxian committed:
Add TimeUtil class

euchangxian committed:
Update Tasks to take in LocalDate

euchangxian committed:
Update LocalDateTime display format

euchangxian committed:
Improve error message for invalid date inputs

euchangxian committed:
Fix TimeUtil::parseDateTimeString

Fix parseDateTimeString() wrongly parsing yyyy-mm-dd, or inputs
containing Date only, as invalid inputs.

euchangxian committed:
Merge branch 'branch-Level-8'

Update Duke to parse date inputs.

euchangxian committed:
Add UiService class.

This class is responsible for the printing of messages to the console
output, and uses OutputService to handle the formatting and indentation
of Strings.

euchangxian committed:
Replace usage of OutputService to UiService

euchangxian committed:
Add TaskList class

Operations on the TaskList like add, delete, mark etc will be handled by
TaskList.

euchangxian committed:
Move Task and related subclasses to package tasks

euchangxian committed:
Update parsing of special strings into dates

"today" and "tomorrow" both return the end of their respective days.
Display formatting is updated to display minutes too.

euchangxian committed:
Merge branch 'branch-A-MoreOOP'

Abstract methods and responsibilities into separate classes for
separation of concern.

euchangxian committed:
Separate classes into respective packages

euchangxian committed:
Merge branch 'branch-A-Packages'

This is to provide a logical grouping for different classes and their
responsibilities, improving readability.

euchangxian committed:
Fix wrong display strings

euchangxian committed:
Remove support for yyyy MM date formats

euchangxian committed:
Add more specific Exceptions

euchangxian committed:
Add abstract Command and subclasses

Abstracts logic of inputs into individual Commands subclasses.

euchangxian committed:
Update parsing logic to support Command Pattern

euchangxian committed:
Remove unused fields and methods

euchangxian committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

euchangxian committed:
Merge branch 'branch-A-Gradle'

Add Gradle support to Duke project.

euchangxian committed:
Add TimeUtil tests

euchangxian committed:
Extract out parsing of input to its own method

This is to allow for a more straightforward unit testing of
CliParserService.

euchangxian committed:
Move responsiblity of trimming input

Move from parse() to parseCommandAndArguments(). This is to allow for
better unit testing as well

euchangxian committed:
Convert parsed commandType to lower case

euchangxian committed:
Add CliParserServiceTest

euchangxian committed:
Merge branch 'branch-A-JUnit'

Add JUnit tests for CliParserService and TimeUtil, as these are critical
service/util classes.

euchangxian committed:
Add JavaDocs to methods and classes

euchangxian committed:
Merge branch 'branch-A-JavaDoc'

Adds JavaDocs to classes and methods. This improves maintainability.

euchangxian committed:
Add checkstyle to Gradle

euchangxian committed:
Reformat code based on checkstyle

euchangxian committed:
Fix checkstyle violations

euchangxian committed:
Merge branch 'branch-A-CodingStandard'

Implement checkstyle and fix checkstyle violations.

euchangxian committed:
Add printFoundTasks in UiService

Refactor printTask in OutputService to formatTaskList for increased
flexibility.

euchangxian committed:
Add find command

find searches the taskList for taskNames containing the given keyword,
then prints the matching tasks.

euchangxian committed:
Merge branch 'branch-Level-9'

Add find command which allows users to search for tasks based on a
keyword.

euchangxian committed:
Add JavaDocs for methods invoked by find command

Fixes CheckStyle violations

LicongHuang commented on src/main/java/duke/service/StorageService.java:
```diff
@@ -0,0 +1,158 @@
+     *
+     * @return {@code true} if the file was corrupted, otherwise {@code false}.
+     */
+    public boolean wasFileCorrupted() {
```
Method name should start with a verb, i.e. checkFileCorrupted()

LicongHuang commented on src/main/java/duke/tasks/TaskList.java:
```diff
@@ -0,0 +1,116 @@
+        return Optional.of(task);
+    }
+
+    public List<Task> searchTasks(String keyword) {
```
Good use of monads

PearlynnT replied:
Second this

LicongHuang commented on text-ui-test/input.txt:
```diff
@@ -0,0 +1,9 @@
+todo master rizz
```
I also wish

LicongHuang reviewed (commented):
Overall very well written, follows the style guide very well, lgtm, can push to production

euchangxian committed:
Merge branch 'branch-A-CheckStyle'

euchangxian committed:
Extract main method from Duke to Main class

This is in preparation to integrate JavaFX with Duke, preventing the
Duke class from becoming too bloated.

euchangxian committed:
Refactor Duke class

euchangxian committed:
Add JavaFX GUI

Refactor existing classes, mainly changing the return type of UiService
to return a String which is displayed in the GUI window.

euchangxian committed:
Merge branch 'branch-Level-10'

Add a GUI to Duke.

PearlynnT commented on src/main/java/duke/service/CommandFactory.java:
```diff
@@ -0,0 +1,110 @@
+                throw new InvalidCommandInputException("Provide a keyword for find command! :<");
+            }
+            return new FindTaskCommand(dukeBot, uiService, args.get(0));
+        case "todo":
```
```suggestion
        case "todo":
        // Fallthrough
```
The explicit // Fallthrough comment should be included whenever there is a case statement without a break statement.

PearlynnT commented on src/main/java/duke/service/CommandFactory.java:
```diff
@@ -0,0 +1,110 @@
+            }
+            return new FindTaskCommand(dukeBot, uiService, args.get(0));
+        case "todo":
+        case "deadline":
```
```suggestion
        case "deadline":
        // Fallthrough
```
The explicit // Fallthrough comment should be included whenever there is a case statement without a break statement.

PearlynnT commented on src/main/java/duke/service/StorageService.java:
```diff
@@ -0,0 +1,158 @@
+
+        try (FileInputStream fis = new FileInputStream(file);
+             ObjectInputStream ois = new ObjectInputStream(fis)) {
+            @SuppressWarnings("unchecked")
```
Appropriate use

PearlynnT reviewed (commented):
LGTM Great job! :smile:

euchangxian committed:
Add assert statements to TaskList

TaskFactory is responsible for the creation of new Tasks, and in no
cases does it return a null task.
Since there should only be one way of adding a task to TaskList, by
calling TaskList::add, there should be no null tasks added to the
TaskList. Neither should there by any null task retrieved by any
exposed methods.

euchangxian committed:
Add final modifier to ParseResult fields

This is to better reflect immutability.

euchangxian committed:
Fix UnmarkTaskCommand calling wrong UiService method

euchangxian committed:
Change formatTaskList to use Streams

euchangxian committed:
Add unit test for formatTaskList

euchangxian committed:
Remove unused import

euchangxian committed:
Reorganize imports in OutputServiceTest

euchangxian committed:
Add java_ci.yml file to ensure code quality

euchangxian committed:
Merge pull request #3 from euchangxian/branch-A-Assertions

Add assert statements to TaskList

euchangxian committed:
Merge pull request #2 from euchangxian/branch-A-Streams

Branch-A-Streams

euchangxian committed:
Merge pull request #1 from euchangxian/branch-A-CI

Add java_ci.yml file to ensure code quality

euchangxian committed:
Fix checkstyle violations

euchangxian committed:
Merge pull request #5 from euchangxian/branch-A-CodeQuality

Improve Code Quality

euchangxian committed:
Add more strings that can be understood as dates

Duke can now parse dates like "monday", "tuesday",..., etc
and understands it as the next Monday, Tuesday etc.

euchangxian committed:
Add test for days of week string input

euchangxian committed:
Merge pull request #6 from euchangxian/branch-C-NaturalDates

Add more special date inputs

euchangxian committed:
Update Event Task date arguments logic

If the start date supplied is after the end date, an exception is
thrown.

euchangxian committed:
Merge branch 'branch-A-MoreErrorHandling'

euchangxian committed:
Add Ui.png

euchangxian committed:
Update README.md

euchangxian committed:
Merge branch 'branch-A-UserGuide'

euchangxian committed:
Add JavaFX dependencies in build.gradle

euchangxian committed:
Update javaFX dependencies
